src/CMakeLists.txt

```diff
@@ -0,0 +1,88 @@
+include(GNUInstallDirs)
+
+set(SMFC_SOURCES
+  smfc/codegen.cc
+  smfc/cpp_generator.cc
+  smfc/go_generator.cc
+  smfc/main.cc)
+
+set(SMFC_LIBS
+  Flatbuffers::flatbuffers
+  Boost::system
+  Boost::filesystem
+  gflags_static
+  glog::glog
+  ${CMAKE_THREAD_LIBS_INIT})
+
+add_executable(smfc ${SMFC_SOURCES})
+target_link_libraries(smfc PRIVATE ${SMFC_LIBS})
+target_include_directories(smfc PRIVATE src/smfc)
+install(TARGETS smfc
+  EXPORT smf-export
+  DESTINATION ${CMAKE_INSTALL_BINDIR})
+
+# TODO: rpc_generated.h should not be checked into git. instead, it should be
+# generated at build-time and installed/packaged explicitly.
+smfc_gen(
+  CPP
+  TARGET_NAME raw_rpc_gen
+  OUTPUT_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/include/smf
+  SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/core/rpc.fbs)
+
+file(GLOB smf_srcs "${CMAKE_CURRENT_SOURCE_DIR}/core/*.cc")
+
+add_library(smf STATIC
+  ${raw_rpc_gen}
+  ${smf_srcs})
+
+set_target_properties(smf PROPERTIES VERSION ${PROJECT_VERSION})
+
+target_include_directories(smf PUBLIC
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>)
+
+target_compile_definitions(smf PUBLIC -DXXH_PRIVATE_API)
+
+target_link_libraries(smf
+  PUBLIC
+  Seastar::seastar
+  Flatbuffers::flatbuffers
+  Zstd::zstd
+  Hdrhistogram::hdr_histogram
+  xxHash::xxhash)
+
+if(CMAKE_BUILD_TYPE MATCHES Debug)
+  target_compile_options(smf PRIVATE
+    -fdiagnostics-color=auto
+    -Wall
+    -Werror
+    -Wextra
+    -Wformat
+    -Wmissing-braces
+    -Wparentheses
+    -Wpointer-arith
+    -Wformat-security
+    -Wunused
+    -Wno-unused-parameter
+    -Wcast-align
+    -Wno-missing-field-initializers
+    -Wdelete-non-virtual-dtor
+    -Wno-ignored-qualifiers)
+endif()
+
+install(TARGETS smf
+  EXPORT smf-export
+  LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+  ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+
+if(SMF_ENABLE_BENCHMARK_TESTS)
+  add_subdirectory(benchmarks)
+endif()
+
+if(SMF_ENABLE_INTEGRATION_TESTS)
+  add_subdirectory(integration_tests)
+endif()
+
+if(SMF_ENABLE_UNIT_TESTS)
+  add_subdirectory(tests)
+endif()
```

Comment:
what about the installation code i.e.:

```
# configuration for downstream projects
if(SMF_INSTALL)
  include(GNUInstallDirs)
  include(CMakePackageConfigHelpers)
  set(install_cmakedir ${CMAKE_INSTALL_LIBDIR}/cmake/smf)
  install(
    DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/src/include/
    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
  install(
    TARGETS
      smfc
    EXPORT smf-export
    DESTINATION ${CMAKE_INSTALL_BINDIR})
  install(
    TARGETS
      smf
    EXPORT smf-export
    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
  install(
    EXPORT smf-export
    FILE smfTargets.cmake
    NAMESPACE smf::
    DESTINATION ${install_cmakedir})
  configure_package_config_file(
    ${CMAKE_CURRENT_LIST_DIR}/cmake/smf-config.cmake.in
    ${CMAKE_CURRENT_BINARY_DIR}/smf-config.cmake
    INSTALL_DESTINATION ${install_cmakedir})
  write_basic_package_version_file(
    ${CMAKE_CURRENT_BINARY_DIR}/smf-config-version.cmake
    VERSION ${PROJECT_VERSION}
    COMPATIBILITY ExactVersion)
  install(
    FILES
    ${CMAKE_CURRENT_BINARY_DIR}/smf-config.cmake
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/FindBoost.cmake
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/smfc_generator.cmake
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/FindZstd.cmake
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/FindxxHash.cmake
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/FindHdrhistogram.cmake
    ${CMAKE_CURRENT_SOURCE_DIR}/cmake/FindFlatbuffers.cmake
    DESTINATION ${install_cmakedir})
  export(
    EXPORT smf-export
    FILE ${CMAKE_CURRENT_BINARY_DIR}/smfTargets.cmake
    NAMESPACE smf::)
  export(PACKAGE smf)
endif()
```

Reply:
it's in the root `CMakeLists.txt` though I have tested it yet. I'll do that when I integrate into the sample project.

Reply:
when i tried, cmake complained that the targets must be declared in the same file which is why i had to move them up one directory.

Reply:
that's true, but I don't think there is anything wrong with adding the `install` directives where the targets are declared.